### ISSUE TITLE
Latest Hammerhead Update (Build Version: 1.509.1960) Broke the Ki2 display function

### ISSUE THREAD

**Issue by 95wildtt** [valterc/ki2#242]
After installing this update, the Ki2 app (8.0) no longer displays on my K2... 
It loads and shows it's connected to my Di2 in the Ki2 sensor list, but no longer appears on the screen at all. 
It worked perfectly, just prior to this latest Karoo update. :-(

Just to verify, I removed Ki2 from the K2, and then re-side loaded it, still no luck...

**Comment by valterc** [comment]
I can confirm the same. The version 1.509.1960 broke quite a lot of things, so if you want to use Ki2 hold on before installing that update for now.

**Comment by JBonitatibus** [comment]
My gear display is working but the overlay and the switches are not.  Also, the preload function was causing the Karoo 2 to boot to a white screen.

**Comment by 95wildtt** [comment]
Now that you mention it, I only use the overlay, which isn't working. 
My switches aren't either.
Good catch.

**Comment by CamJones64** [comment]
At the end of my group ride this morning, the update was installed (Build 1.509.1960.12) and lost the overlay and the operation of the hood switches. One of the other group members also has same issue. 

Hoping this is a simple fix, as seems like Hammerhead could start losing a few owners pretty quickly if we cannot display simple data just because some parts are not aligned to their brand. Quick way to destroy value.

**Comment by markuuu** [comment]
Copy of my post to the HH thread on this issue (https://support.hammerhead.io/hc/en-us/community/posts/29781899128091-white-screen-after-update-1-509-1960?page=1#community_comment_29826016125083) 

I am also experiencing this issue, and I believe this highlights a couple of areas where Hammerhead could improve the Karoo ecosystem:

1)  Allow users to uninstall the most current release / install current -1 release, so we can regain functionality when a new release breaks something (be it HH breaking something in their own code or a third party app)

2)  Include Ki2 in HH's beta testing process before releasing an update.  I know, I know, Ki2 is a third party app and HH can't afford to verify compatibility of all third party apps on each release.   I submit that since the Ki2 app provides a feature set that HH is unable to, and is a core feature set for a large enough portion of HH's customer base that breaking it has a significant impact on user/customer satisfaction and future upgrade decisions, that including it in the beta testing process would be in HH's best interest.  Users are more likely to spend hundreds of dollars on a different GPS to regain shifting / switch integration than to spend thousands of dollars to switch all their Di2 drivetrains to SRAM to do so.

**Comment by ConstantVanWyk** [comment]
I am discombobulated that HH does not include this in their regression testing. Speaks volumes for their overall professionalism.

**Comment by CamJones64** [comment]
Looking at the switch issue I noticed the following. Setup is for D-Fly in the Shimano system, and is set up for swapping screens in Ki2 setup, but shows as no action in the Ki2 devices screen.
![Screenshot_20241006-110107](https://github.com/user-attachments/assets/edf020a5-40ac-47be-8c2e-c124349fc95c)
![PXL_20241006_030206243](https://github.com/user-attachments/assets/b928217c-08a9-4cb1-b8e7-003af257b8f3)
![PXL_20241006_030214377](https://github.com/user-attachments/assets/aa8b57fd-5f14-4fd6-99e9-eceeef3767ac)
![PXL_20241006_030239694](https://github.com/user-attachments/assets/1a26380d-5799-4df9-b3bd-a88e6712aa9b)

**Comment by EddyS33** [comment]
@CamJones64 - re your last photo, the display will show the relevant action for the channel in real-time when you press the button.

**Comment by proseLA** [comment]
> 1. Allow users to uninstall the most current release / install current -1 release, so we can regain functionality when a new release breaks something (be it HH breaking something in their own code or a third party app)

that will never happen.  i was on HH about this many many moons ago...

> 2. Include Ki2 in HH's beta testing process before releasing an update.  I know, I know, Ki2 is a third party app and HH can't afford to verify compatibility of all third party apps on each release.   I submit that since the Ki2 app provides a feature set that HH is unable to, and is a core feature set for a large enough portion of HH's customer base that breaking it has a significant impact on user/customer satisfaction and future upgrade decisions, that including it in the beta testing process would be in HH's best interest.  Users are more likely to spend hundreds of dollars on a different GPS to regain shifting / switch integration than to spend thousands of dollars to switch all their Di2 drivetrains to SRAM to do so.

i think complaining about HH on this ki2 repo is mis-guided.  if shimano wanted to shut this app down the same way they shut down HH, rest assured they would.  

i can not speak as to why the latest version of the HH firmware broke this app; i am not java fluent.  but this app is a plugin; and perhaps this plugin is the one that needs to conform to what the new firmware now does.  like many others here, i am a huge fan of this app and of all the work @valterc does here.  but i think only he can state whether the problem is the HH firmware or his plugin.  as he stated above, hold off for now on doing this update.  for those who have already updated, i feel your pain.

> I am discombobulated that HH does not include this in their regression testing. Speaks volumes for their overall professionalism.
> 
> As to the 3rd party app, argument. I offer the perspective that there is a significant user base using this app (due to Shimano & SRAM's greed) that it can be argued that it is good practice and sensible to at least regression test that a widely used app does not break.

yeah, i disagree.  i find HH to be very professional as well as responsive, when they can be.  software always has bugs, and when HH breaks something, i find them to be responsive to putting out a new version rather quickly.  and again, why they would need to ascertain whether a plugins code conforms to their SDK, and adjust their code accordingly, you got me.  everyone wants to say HH broke my Ki2 app; i just think the verdict is still out...

with regards to the specific situation that we are now in; i can see Shimano's greed, but not SRAM.  shimano are the ones that pulled the functionality; despite allowing all other bike computers to connect to their drivetrain.  so i'm not sure how this is SRAMs greed.  and until you can provide real data that suggests a significant user base out there using this app other than your empirical findings, i am hesitant to agree with you.

that said, i am far more likely to buy a SRAM equipped bike for my n+1 (due to shimano's greed); and i think we are better suited on this repo to speak specifically about ki2 than all of the other stuff.  which i will now do as i speak...

**Comment by RustedRoot** [comment]
Agreed on Shimano... 

You're aware the K3, or whatever it's called, does not have Ki2 functionality? HH has yet to release its SDK, and without it, no Ki2, this no Di2. This has been going on since the K3's release. 

SRAM's electronic sets connect just fine. It's justified to question SRAM's motives now;  not looking like the innocent party anymore IMO.

**Comment by valterc** [comment]
We are a couple months away from black Friday deals, at this point I would start to recommend a Garmin unit. 

While hammerhead should not hold back their development to support third-party apps, the lack of any engagement with the community and the constant cat and mouse game of development just to show a number on screen and make beeps is getting old. Note that this update not only breaks Ki2, but also other SDK apps. 

The SDK has never been updated since it was released in 2021, it is severely limited and full of bugs that have been reported but never fixed.

**Comment by ConstantVanWyk** [comment]
> We are a couple months away from black Friday deals, at this point I would start to recommend a Garmin unit.
> 
> While hammerhead should not hold back their development to support third-party apps, the lack of any engagement with the community and the constant cat and mouse game of development just to show a number on screen and make beeps is getting old. Note that this update not only breaks Ki2, but also other SDK apps.
> 
> The SDK has never been updated since it was released in 2021, it is severely limited and full of bugs that have been reported but never fixed.

Really appreciate your time and effort to keep this going, mate. 

Not the forum for it but I came from Garmin back in the day, my head is starting to lean towards Wahoo.

**Comment by benjaminhendler** [comment]
Hi, what possibilities do you think we have fixing this? Ki2 is the only reason I’m holding onto the Karoo at the moment. I donated another couple of coffees just to show my appreciation.

**Comment by Mputxeta** [comment]
valterc, thank for your work.
I cannot believe HH. I don't want they adding Di2 features, but not cutting off Ki2 or any other 3rd party apps. It's android!!.
Do you think you can fix in a next realease de Ki2 working on the HH Karoo firmware? Or just go for a Garmin unit a stop waiting on this guys.

thanks for your amazing work, it's been useful an fun so far...

**Comment by valterc** [comment]
> Hi, what possibilities do you think we have fixing this? Ki2 is the only reason I’m holding onto the Karoo at the moment. I donated another couple of coffees just to show my appreciation.

Thanks Ben! And thanks everyone for the kind words. It's appreciated 🙂

There is a fix (#243) already done for the blank screen/display problems. There is also another fix (#244) in the works for the audio alerts. I still need to go out and test the fixes in a couple of rides, probably will be released by the end of the week.

**Comment by markuuu** [comment]
Thanks for working on fixing this so quickly, Valterc!  Enjoy the coffee.  :-)

**Comment by 95wildtt** [comment]
> > Hi, what possibilities do you think we have fixing this? Ki2 is the only reason I’m holding onto the Karoo at the moment. I donated another couple of coffees just to show my appreciation.
> 
> Thanks Ben! And thanks everyone for the kind words. It's appreciated 🙂
> 
> There is a fix (#243) already done for the blank screen/display problems. There is also another fix (#244) in the works for the audio alerts. I still need to go out and test the fixes in a couple of rides, probably will be released by the end of the week.

WOW!
That's fast... Thank you valterc.

**Comment by e17biker** [comment]
> > Hi, what possibilities do you think we have fixing this? Ki2 is the only reason I’m holding onto the Karoo at the moment. I donated another couple of coffees just to show my appreciation.
> 
> Thanks Ben! And thanks everyone for the kind words. It's appreciated 🙂
> 
> There is a fix (#243) already done for the blank screen/display problems. There is also another fix (#244) in the works for the audio alerts. I still need to go out and test the fixes in a couple of rides, probably will be released by the end of the week.

Thanks for all your efforts it's really appreciated. Enjoy the coffee.

**Comment by CamJones64** [comment]
Appreciate your efforts on this @valterc. I am also raising with HH as a safety issue regarding the hood switches, as now need to take hand off the bars. Well, got to try any angle we can to get them to respond.

**Comment by coastbum** [comment]
I applied #243 and still get the white screen on turnup.   If I turn off pre-load it prevents white screen but none of the DI2 sensors get loaded on displays until I open up KI2 and fish around and then sometimes that will add the DI2 stuff to screen.   Unfortunately, I'm beginning to think going back to Garmin is in the stars.   HH folks seem to have lost their way and combined with the basic "Japanese bicycle vendor" craziness it is getting to be a bore.   THANK YOU for all your wonderful work on KI2 !!!   I can't tell you how much your efforts are appreciated.    Off to buy you a coffee.

**Comment by valterc** [comment]
Thanks everyone!

Ki2 version [8.1](https://github.com/valterc/ki2/releases/tag/8.1) is now out with a fix for the problems with the latest Karoo release. You can update directly from the Karoo, or download the APK file (https://github.com/valterc/ki2/releases/tag/8.1) and install via a computer.

Let me know if that version works or if you have problems.

**Comment by CamJones64** [comment]
Awesome news @valterc, have just got home so will give it a try shortly. Appreciate all your efforts.

**Comment by paulus87** [comment]
@valterc Thank you very much for all your good work. Bought you a coffee. A month ago I almost wanted to quit Karoo and bought a Wahoo, but now everything works again! (No freezes or blank screen anymore)

**Comment by GainfulShrimp** [comment]
I want to add my own sincere appreciation for all your efforts @valterc! I was so disappointed when my Karoo2 started misbehaving following the recent update... 
The relief I felt when you updated Ki2 to fix the issues so quickly was well worth the price of a coffee or three; so I've just donated to show my thanks. ☕ ☕ ☕ 
Keep up the great work and thanks again!

**Comment by pictoruws** [comment]
Thank you @valterc for you speedy fix for this issue.  I just sent you a coffee donation. Cheers.

**Comment by valterc** [comment]
Thanks everyone! 🙂

**Comment by ConstantVanWyk** [comment]
Coffee incoming as a token of appreciation.

![image](https://github.com/user-attachments/assets/54d8c319-39ff-492d-8b4e-4ab559cae3d6)

**Comment by 95wildtt** [comment]
More coffee on the way... Thank you Valterc.
Quick question:
- When using the Gear/Shift field on my K2, is the small battery symbol dynamic (does it change as the Di2 battery is used) and will the battery symbol turn red when the Di2 battery is low (like at 10%) or will it stay green until it's dead?

**Comment by JayUK** [comment]
Just curious..... How did you manage to get this working? HH haven't updated the SDK (the changes they made in the release were internal to their software?) So how did you know what you needed to change?

Thanks,
J.
